### PR TITLE
Merge 2.9 to 3.0

### PR DESCRIPTION
**Pull request by hpidcock**
Forward ports from 2.9 + backport from 3.1 for pebble/plan + fix unused linter errors.

Back ports:
- #15234

Forward ports:
- #15264
- #15274
- #15260
- #15272
- #15300
- #15312
- #15309
- #15317
- #15320
- #15332
- #15335
- #15334
- #15337
- #15344

Conflicts:
- api/client/imagemanager/client_test.go
- apiserver/facades/agent/caasapplication/mock_test.go
- apiserver/facades/agent/uniter/uniter.go
- apiserver/facades/client/application/mock_test.go
- apiserver/facades/client/charms/client_test.go
- apiserver/facades/client/client/client_test.go
- apiserver/facades/client/cloud/cloudV2_test.go
- apiserver/facades/client/modelmanager/shims.go
- apiserver/facades/client/storage/storage.go
- apiserver/facades/controller/remoterelations/mock_test.go
- apiserver/mocks/resources_mock.go
- apiserver/resources_mig_test.go
- caas/kubernetes/provider/application/application.go
- charmhub/refreshconfig.go
- cmd/containeragent/main_nix.go
- cmd/juju/action/package_test.go
- cmd/juju/action/run.go
- cmd/juju/action/showoutput.go
- cmd/juju/application/bundle_test.go
- cmd/juju/application/deployer/bundlehandler_test.go
- cmd/juju/application/removeapplication.go
- cmd/juju/application/removeunit.go
- cmd/juju/application/resolved_test.go
- cmd/juju/application/scaleapplication_test.go
- cmd/juju/backups/backups.go
- cmd/juju/caas/update_test.go
- cmd/juju/charmhub/download.go
- cmd/juju/charmhub/download_test.go
- cmd/juju/commands/bootstrap_test.go
- cmd/juju/commands/exec_test.go
- cmd/juju/gui/gui.go
- cmd/juju/gui/gui_test.go
- cmd/juju/model/exportbundle.go
- cmd/juju/romulus/sla/sla_test.go
- cmd/jujud/agent/deploy_test.go
- go.mod
- go.sum
- provider/azure/storage.go
- provider/maas/maas2_test.go
- provider/openstack/config_test.go
- service/agentconf_test.go
- state/backups/restore_test.go
- state/upgrades_test.go
- state/user.go
- state/watcher/txnwatcher_test.go
- worker/httpserver/cert_test.go
- worker/proxyupdater/proxyupdater_test.go

### CODE REVIEW

**Review by wallyworld — APPROVED** [review]

File: cmd/juju/charmhub/download_test.go
Comment (hpidcock) [inline cmd/juju/charmhub/download_test.go:1]: @SimonRichardson can you please review this file.

File: cmd/juju/charmhub/download.go
Comment (hpidcock) [inline cmd/juju/charmhub/download.go:1]: @SimonRichardson can you please review this file.

File: cmd/juju/charmhub/download_test.go [hunk 1]
```diff
@@ -136,9 +136,32 @@ func (s *downloadSuite) TestRunWithCustomCharmHubURL(c *gc.C) {
...
+
+	ctx := commandContextForTest(c)
+	err = command.Run(ctx)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *downloadSuite) TestRunWithUnsupportedSeriesReturnsSecondAttempt(c *gc.C) {
+	defer s.setUpMocks(c).Finish()
+
+	s.expectRefreshUnsupportedSeries()
```
Comment (hpidcock) [inline cmd/juju/charmhub/download_test.go:164]: @SimonRichardson this merged test fails on
```
c.Assert(err, gc.ErrorMatches, `"test" does not support series "jammy" in channel "stable".  Supported series are: bionic, trusty, xenial.`)
```

I'm not sure why we are returning a different error now.

```
13:56:45 FAIL: download_test.go:161: downloadSuite.TestRunWithUnsupportedSeriesReturnsSecondAttempt
13:56:45 
13:56:45 download_test.go:176:
13:56:45     c.Assert(err, gc.ErrorMatches, `"test" does not support series "jammy" in channel "stable".  Supported series are: bionic, trusty, xenial.`)
13:56:45 ... error string = "\"test\" does not support series \"bionic\" in channel \"stable\".  Supported series are: bionic, trusty, xenial."
13:56:45 ... regex string = "\"test\" does not support series \"jammy\" in channel \"stable\".  Supported series are: bionic, trusty, xenial."
```

**Pull request merged**